cli/src/helpers/mpl-token-auth-rules.ts

```diff
@@ -1,79 +1,159 @@
+  tx.recentBlockhash = blockhash;
+  tx.feePayer = payer.publicKey;
+  const sig = await connection.sendTransaction(tx, [payer]);
+  await connection.confirmTransaction(sig);
+
+  return ruleSetAddress;
+};
+
+export const createOrUpdateLargeRuleset = async (
+  connection: Connection,
+  payer: Keypair,
+  name: string,
+  data: Uint8Array,
+) => {
+  if (data.length <= CHUNK_SIZE) {
+    return createOrUpdateRuleset(connection, payer, name, data);
+  }
+
+  const chunks = Math.ceil(data.length / CHUNK_SIZE);
+
+  for (let i = 0; i < chunks; i++) {
+    const chunk = data.slice(i * CHUNK_SIZE, Math.min((i + 1) * CHUNK_SIZE, data.length));
+    console.log(`   + writing data slice ${i + 1} of ${chunks}: ${chunk.length} bytes`);
+    await writeAndPuff(connection, payer, name, chunk, i == 0);
+  }
+
+  const [bufferAddress] = await findRuleSetBufferPDA(payer.publicKey);
+  return createOrUpdateRuleset(connection, payer, name, bufferAddress);
+};
+
+export const writeAndPuff = async (
+  connection: Connection,
+  payer: Keypair,
+  name: string,
+  data: Uint8Array,
+  overwrite = false,
 ) => {
-    const ruleSetAddress = await findRuleSetPDA(payer.publicKey, name);
-
-    let createIX = createCreateOrUpdateInstruction(
-        {
-            payer: payer.publicKey,
-            ruleSetPda: ruleSetAddress[0],
-            systemProgram: SystemProgram.programId,
-        },
-        {
-            createOrUpdateArgs: {__kind: "V1", serializedRuleSet: data },
-        },
-        PROGRAM_ID,
-    )
-
-    const tx = new Transaction().add(createIX);
-
-    const { blockhash } = await connection.getLatestBlockhash();
-    tx.recentBlockhash = blockhash;
-    tx.feePayer = payer.publicKey;
-    const sig = await connection.sendTransaction(tx, [payer], { skipPreflight: true });
-    await connection.confirmTransaction(sig, "finalized");
-    return ruleSetAddress[0];
-}
+  const bufferAddress = await findRuleSetBufferPDA(payer.publicKey);
+
+  const writeIX = createWriteToBufferInstruction(
+    {
+      payer: payer.publicKey,
+      bufferPda: bufferAddress[0],
+      systemProgram: SystemProgram.programId,
+    },
+    {
+      writeToBufferArgs: { __kind: 'V1', serializedRuleSet: data, overwrite },
+    },
+    PROGRAM_ID,
+  );
+
+  const ruleSetAddress = await findRuleSetPDA(payer.publicKey, name);
+
+  const puffIX = createPuffRuleSetInstruction(
+    {
+      payer: payer.publicKey,
+      ruleSetPda: ruleSetAddress[0],
+      systemProgram: SystemProgram.programId,
+    },
+    {
+      puffRuleSetArgs: { __kind: 'V1', ruleSetName: name },
+    },
+    PROGRAM_ID,
+  );
+
+  const tx = new Transaction().add(writeIX, puffIX);
+
+  const { blockhash } = await connection.getLatestBlockhash();
+  tx.recentBlockhash = blockhash;
+  tx.feePayer = payer.publicKey;
+  const sig = await connection.sendTransaction(tx, [payer]);
+  await connection.confirmTransaction(sig);
+  return bufferAddress[0];
+};
 
 export const validateOperation = async (
```

Comment:
I guess we leave `validateOperation` in case we want to use it for a CLI validate in the future?